### ISSUE TITLE
TF2 in-game chat font issue

### ISSUE THREAD

**Issue by jorgenpt** [ValveSoftware/Source-1-Games#67]
Issue transferred from ValveSoftware/steam-for-linux#578
@artixlight posted at 2013-01-07T05:11:45Z:

When typing text in-game for TF2, the fonts looked squished and it was hard to read. On searching the internet, I found this tutorial: http://www.dodbits.com/dods/index.php/source-graphics/source-gui-hud-tutorials/19-chat-area-in-hl2-games

Browsing through the _SteamApps_ directory, I copied the file **Multiplayer OB Linux Binaries.gcf** to a Windows machine and used _GCFScape_ to extract **ChatScheme.res** file.

When the .res file was opened, under **FONTS** section, I found that it uses Verdana fonts. This font comes pre-installed on Windows machine but not on Linux distributions as it was optional and was copyrighted by Microsoft. One could install the required fonts in order for the chat text to appear as it would in Windows version of TF2. This issue was previously reported in steamcommunity.com: http://steamcommunity.com/app/221410/discussions/0/846939071403952294/

Attached is the comparison between 2 different fonts rendering for the in-game chat system. Open both in full resolutions and compare them side-by-side:
- Installed Verdana font in my Linux distribution. Click on the image to see the full resolution. Text is clearly legible.
  ![2013-01-07_00003](https://f.cloud.github.com/assets/3202292/46630/2ae81f18-587e-11e2-944b-97ad3eecaee6.jpg)
- Without Verdana font installed in my Linux distribution. Click on the image to see the full resolution. Text appears to be squished.
  ![2013-01-07_00005](https://f.cloud.github.com/assets/3202292/46634/91e81cf4-587e-11e2-93cf-639fd7011a76.jpg)
### Steps to reproduce the issue:
1. Install Verdana font. There are ways to install it from repositories. Refer to distribution's help or search around the internet.
2. Run TF2 and load a map. Type some text in the chat. Take a screenshot (F12 key by default).
3. Exit TF2.
4. Uninstall Verdana font and run TF2 again.
5. Load a map and type some text in the chat. Take a screenshot.
6. Compare both screenshots.
### Suggested solutions:
1. Modify **ChatSheme.res** file such that it can fallback to default fonts used by Linux distributions if Verdana was not installed. Downside is that different Linux distributions will come with different default fonts used, which may create additional problems if style consistency was to be maintained.
2. Bundle some fonts into the game that can be redistributed under some sort of license (preferably open source) to maintain consistency. Downside is that some fonts may not cover the whole Unicode range if characters other than Latin alphabets were used.
3. Provide an option in-game for the user to select the font to use, like an option found in most word processing software (e.g. Microsoft Word, LibreOffice Writer). Downside, more coding and possibly new bugs can be introduced.
### System information:

```
Processor Information:
    Vendor:  GenuineIntel
    Speed: 2534 Mhz
    4 logical processors
    3 physical processors
    HyperThreading:  Supported
    FCMOV:  Supported
    SSE2:  Supported
    SSE3:  Supported
    SSSE3:  Supported
    SSE4a:  Unsupported
    SSE41:  Supported
    SSE42:  Supported

Network Information:
    Network Speed:  

Operating System Version:
    Ubuntu 12.04.1 LTS (64 bit)
    Kernel Name:  Linux
    Kernel Version:  3.2.0-34-generic
    X Server vendor:  The X.Org Foundation
    X Server release:  11103000

Video Card:
    Driver:  ATI Technologies Inc. ATI Mobility Radeon HD 5400 Series 

    Driver Version:  4.2.11931 Compatibility Profile Context
    Desktop Color Depth: 24 bits per pixel
    Monitor Refresh Rate: 60 Hz
    VendorID:  0x1002
    DeviceID:  0x68e0
    Number of Monitors:  1
    Number of Logical Video Cards:  1
    Primary Display Resolution:  1366 x 768
    Desktop Resolution: 1366 x 768
    Primary Display Size: 13.54" x 7.60"  (15.51" diag)
                                            34.4cm x 19.3cm  (39.4cm diag)
    Primary VRAM Not Detected

Sound card:
    Audio device: IDT 92HD73C1X5

Memory:
    RAM:  2871 Mb

Miscellaneous:
    UI Language:  English
    LANG:  en_GB.UTF-8
    Microphone:  Not set
    Total Hard Disk Space Available:  27287 Mb
    Largest Free Hard Disk Block:  5393 Mb

Installed software:

Recent Failure Reports:
```

**Comment by kisak-valve** [comment]
Hello, is this still an issue on up to date systems?

**Comment by bootkiller** [comment]
Yes, just tested on CS:Source. Text is completely unreadable under certain situations, like the floor on de_dust2.

**Comment by Staysys** [comment]
Just thought I'd mention here that someone found a workaround:
http://steamcommunity.com/app/221410/discussions/0/618457398964472446/?ctp=2
I'll quote the specific post here:

> For CSS I just copied the font from my Windows install and configured it to be used. Now the chat font in Linux looks the same as Windows and is no longer difficult to read.
>
>Copied vgasys.fon from C:\Windows\Fonts to $home/.steam/steam/steamapps/common/Counter-Strike Source/hl2/resource .
>
>Edited chatscheme.res in the same directory.
>
>Under "ChatFont", I located the setting for the yres (vertical resolution) range I'm using (768 to 1023 or subsection "3" in my case). Changed the "name" from "Verdana" to "System".
>
>Under "CustomFontFiles", I added a new font, "3", set to "resource/vgasys.fon". 

I also got it to work on CS1.6, but I can't remember if I did anything to config files. I have "vgasys.fon" in my cstrike/resource directory though.

**Comment by feikname** [comment]
@kisak-valve This issue still persists, at least in my *Elementary OS 0.4 freya*.

I have microsoft fonts installed.

**Comment by stevethepocket** [comment]
Regardless of what's causing this, it's clear that the game renders any fonts that are bundled with the game just fine. My recommendation is to find a good, easy-to-read, generic font that's either free to use or that you can justify licensing, and retool the game to always use that instead. The game has changed so much in 10 years that changing one of its three fonts won't matter to anyone.

**Comment by stevethepocket** [comment]
No, it's definitely an issue with the game. I think maybe it's unable to access /usr/share/fonts/ or something, because thus far we've seen no evidence of a Source game correctly accessing a font not embedded into the game files somewhere.

**Comment by leewdch** [comment]
what about launching thw game with sudo? or changing /usr/share/fonts permissions?

**Comment by etkaar** [comment]
Did anyone find a solution?

**Comment by lebrunregis** [comment]
Installing/writing a custom hud that change the font/size of the font would probably fix the issue temporarily.

**Comment by etkaar** [comment]
Unfortunately different fonts are totally ignored in the .res files.

**Comment by Neubulae** [comment]
As to this day, this problem still persists, tested on Arch Linux 5.11.7-arch-1-1

**Comment by Bugaboo2000** [comment]
> Did anyone find a solution?

I installed the ttf-ms-fonts and solved for me 
![20210603091814_1](https://user-images.githubusercontent.com/76823368/120644544-d6c9b000-c44d-11eb-8d4c-1fded2857aef.jpg)

**Comment by etkaar** [comment]
> > Did anyone find a solution?
> 
> I installed the ttf-ms-fonts and solved for me
> ![20210603091814_1](https://user-images.githubusercontent.com/76823368/120644544-d6c9b000-c44d-11eb-8d4c-1fded2857aef.jpg)

Really? 😳 Where did you download them?

**Comment by Bugaboo2000** [comment]
> > Replying to [#67 (comment)](https://github.com/ValveSoftware/Source-1-Games/issues/67#issuecomment-853830017)
> 
> Really? flushed Where did you download them?

I downloaded from the official repositories, but in my case i'm using Manjaro. I don't know if links are allowed here, but with a quick Google search you can find how to install on Ubuntu and derivatives

**Comment by lflo5727** [comment]
Going to update here, that this issue still exists (tested on many distros, currently on EndeavourOS (Arch)). Installing the proper Microsoft fonts doesn't seem to do anything for me. It makes text in HUDs pretty difficult to read

**Comment by etkaar** [comment]
Yes, still exists of course. Think they'll fix it twenty years after I was already hundred years dead.

**Comment by lolsniperftw** [comment]
Are there any updates on that issue? Even the game gets a Playable rating because of that on the Deck.

> Some in-game text is small and may be difficult to read

**Comment by spacelike** [comment]
> > Did anyone find a solution?
> 
> I installed the ttf-ms-fonts and solved for me ![20210603091814_1](https://user-images.githubusercontent.com/76823368/120644544-d6c9b000-c44d-11eb-8d4c-1fded2857aef.jpg)

But why did you have to position the camera so that the text is over one of the darkest objects on that entire map?
Is it because if you put the yellow text over any other texture on that yellow map it's difficult to read? That's the entire problem.
On windows you can easily read the text no matter what background it appears over.

I think the problem here is 90% about the shadow behind the text. In the comparison image from @Staysys it's clear that the windows text shadow is solid (there are no transparent parts of the shadow, it's all solid black), while the shadow on linux is not solid black in many places like if transparency is enabled.

If Valve would just disable the transparency from the text shadow it would probably make it MUCH more readable, like on windows.
Then the remaining 10% of the problem is the font rendering itself, which looks much thinner and smaller than the windows one. But that's less of an issue if I can at least read the text with a proper shadow.

**Comment by BenA0** [comment]
Adding to https://github.com/ValveSoftware/Source-1-Games/issues/67#issuecomment-18576014, It seems like the limitation is not the OS itself.

| Native | Native + [font patch](https://steamcommunity.com/sharedfiles/filedetails/?id=2829864534) | Proton |
|--------|--------|--------|
| ![20230815_18h41m29s_grim](https://github.com/ValveSoftware/Source-1-Games/assets/59608414/1dddc5f1-da15-424e-80b8-14554b988786) | ![20230815_18h40m37s_grim](https://github.com/ValveSoftware/Source-1-Games/assets/59608414/ee1c4a32-00dd-40f4-b35f-4fc628084cfb) | ![20230815_18h41m16s_grim](https://github.com/ValveSoftware/Source-1-Games/assets/59608414/68667fd1-2514-4efa-81fb-06192cba8b27) |

I haven't yet checked what is the difference between Proton and native regarding this issue. if anyone is willing to find out it'll be appreciated; due to VAC not working under Proton this "solution" is rather useless.

**Comment by Toni500github** [comment]
I was having the same issue with console and community server browser, the text was small and unreadble, and was happening on all source games. 
Before I switched to EndavourOS from LinuxMint 20.3 I backup everything including fonts, so I found out the problem was with some Microsoft fonts that I didn't had on EndeavourOS, and as mentioned by this [comment](https://github.com/ValveSoftware/Source-1-Games/issues/67#issuecomment-385187967_).

The probable solution for Debian/Ubuntu is to install `ttf-mscorefonts-installer` (or https://launchpad.net/ubuntu/+source/msttcorefonts for Ubuntu)
for arch users try to install the AUR package `ttf-ms-fonts` [Flagged out-of-date (2023-10-29)](https://aur.archlinux.org/pkgbase/ttf-ms-fonts/flag-comment)

For others or if it didn't work, follow this guide http://corefonts.sourceforge.net/

**Comment by purplebar0** [comment]
It seems like *all* font issues have been fixed in the x64 beta. I tested without any UI tweaks or replacing WenQuanYiZenHei (hard-coded last-resort fallback font on Linux), all of these seem to be fixed:
- Font weights appear to work? (The chat is definitely bold now)
  - The last-resort fallback font, whatever it may be now, no longer results in text potentially thinner than the rest of the line
- Console font is now a proper monospace font

**Comment by stevethepocket** [comment]
How old is that beta? This is the first I've heard of it, and to my knowledge it's the only Source 1 game to have a 64 bit version publicly available since they scrapped the one for _Half-Life 2_ over a decade ago.

**Comment by stevethepocket** [comment]
Neat. And now I'm just realizing I don't actually know if the DXVK updates to other games were 64-bit or not. I feel like there would have been some information to that effect if they were.

**Comment by spacelike** [comment]
Although the text definitely looks a LOT better in the new 64-bit version, the text shadow is still not rendering correctly as shown here: https://github.com/ValveSoftware/Source-1-Games/issues/67#issuecomment-18576014

Granted even without a proper shadow the text is much more readable now, but the shadow still helps in certain situations like when the chat window is over a bright texture.

**Comment by Tiagoquix** [comment]
Can confirm the text shadow doesn't render correctly on Linux. On Windows, it looks fine.

**Comment by Tiagoquix** [comment]
This has been fixed as of today's update. Font rendering is now consistent between platforms (change from HL2 Anniversary update).

Example screenshot of chat:

![image](https://github.com/user-attachments/assets/ca3019cc-19e9-4a19-885d-5767cfb98c8f)

**Comment by etkaar** [comment]
😮

**Comment by ghost** [comment]
Legacy runtime 1.0:

![Image](https://github.com/user-attachments/assets/83719960-487a-45f8-9021-6a9dcab03620)

Proton experimental:

![Image](https://github.com/user-attachments/assets/97e1d535-c31e-49a7-9fc2-9d68ec846477)

WIll retest once #7010 is fixed, not sure what caused Proton regression compared to 2023 https://github.com/ValveSoftware/Source-1-Games/issues/67#issuecomment-1679204623

**Comment by Tiagoquix** [comment]
@BenA0 New games should be using the Sniper runtime (3.0) when #7010 is fixed.

**Comment by ghost** [comment]
Native runtime/container:

![Image](https://github.com/user-attachments/assets/40e41c02-c9ec-46c2-be2f-68f3cd7e6e61)

**Comment by Tiagoquix** [comment]
> Although the text definitely looks a LOT better in the new 64-bit version, the text shadow is still not rendering correctly as shown here: [#67 (comment)](https://github.com/ValveSoftware/Source-1-Games/issues/67#issuecomment-18576014)
> 
> Granted even without a proper shadow the text is much more readable now, but the shadow still helps in certain situations like when the chat window is over a bright texture.
> 
> EDIT: Adding a new pic to show what I'm talking about ![tf2-chat-text](https://private-user-images.githubusercontent.com/3375079/324226704-9dc6740d-1c1f-4022-ac07-4d1b4c7c9caa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDAxOTYwODYsIm5iZiI6MTc0MDE5NTc4NiwicGF0aCI6Ii8zMzc1MDc5LzMyNDIyNjcwNC05ZGM2NzQwZC0xYzFmLTQwMjItYWMwNy00ZDFiNGM3YzljYWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMjJUMDM0MzA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjBiYjU5NGNiNWQ5NGQ2Y2IxZDA5YjY1NGY1ZGI3YzgwZDE1MTYxNTZlMmI0ODA3YmJiYjk5OWVhOWQ4N2MzMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FYpeDwcC5-bf8SE3WlMQTaRW5tQPAPuHRvesZgJClPE) Notice how the shadow almost seems to not touch the text, there's a gap between the text and the shadow. Almost like it's 1 pixel off. Also the shadow varies in intensity a lot. On windows it's a solid black without any other shades of gray.

The issue mentioned by that comment still happens. Font consistency between platforms is somewhat OK right now; however, the font blends in with the environment that is behind it as the background on Linux. This does not happen on Windows, as the shadows are correctly displayed.

In other words, reading text chat is harder when the chat has a light background (due to the map's textures). On Windows, this doesn't happen because the shadows display correctly and thus highlight the fonts to make them more readable independently of the text chat current map's background.

Related: https://github.com/ValveSoftware/Source-1-Games/issues/1883

**Comment by Tiagoquix** [comment]
@kisak-valve the main issue here has since been fixed (fonts looking to small), but the problem  that remains is the font shadow (as pointed out in recent comments on this issue). should this one be closed and a new one created just for the shadow issue? thanks!

**Comment by Tiagoquix** [comment]
also, there are a lot of open issues regarding fonts on Linux on Source 1 games, so I think a megathread could be made to unify many of them since they mostly focus on the same problems (such as problems with Asian and Russian fonts)
